libmuscle/python/libmuscle/manager/deadlock_detector.py

```diff
@@ -0,0 +1,220 @@
+                _logger.error(
+                    "Instance %s was waiting for %s, not for %s.",
+                    instance_id,
+                    self._waiting_instances[instance_id],
+                    peer_instance_id)
+            elif self._waiting_instance_ports[instance_id] != (port_name, slot):
+                _logger.error(
+                    "Instance %s was waiting on port[slot] %s[%s], not on %s[%s]",
+                    instance_id,
+                    *self._waiting_instance_ports[instance_id],
+                    port_name, slot)
+            else:
+                del self._waiting_instances[instance_id]
+                del self._waiting_instance_ports[instance_id]
+
+                # Check if we were part of a deadlock
+                for i, instance_list in enumerate(self._detected_deadlocks):
+                    if instance_id in instance_list:
+                        del self._detected_deadlocks[i]
+                        break
+                if not self._detected_deadlocks:
+                    # There are no deadlocks anymore: cancel shutdown
+                    self._shutdown_time = None
+
+    def _check_for_deadlock(self, instance_id: str) -> None:
+        """Check if there is a cycle of waiting instances that involves this instance.
+        """
+        deadlock_instances = [instance_id]
+        cur_instance = instance_id
+        while cur_instance in self._waiting_instances:
+            cur_instance = self._waiting_instances[cur_instance]
+            if cur_instance == instance_id:
+                self._handle_potential_deadlock(deadlock_instances)
+                return
+            deadlock_instances.append(cur_instance)
+        _logger.debug("No deadlock detected")
+
+    def _handle_potential_deadlock(self, deadlock_instances: List[str]) -> None:
+        """Handle a potential deadlock.
+
+        Args:
+            deadlock_instances: list of instances waiting on eachother
+        """
+        shutdown_delay = self._wait_before_shutdown
+        if self._shutdown_time is not None:
+            # Get time until shutdown
+            shutdown_delay = self._shutdown_time - time.monotonic()
+        _logger.fatal(
+                "Potential deadlock detected, aborting run in %d seconds.\n%s",
+                shutdown_delay,
+                self._format_deadlock(deadlock_instances),
+                )
+
+        self._detected_deadlocks.append(deadlock_instances)
+        if self._shutdown_time is None:
+            self._shutdown_time = time.monotonic() + self._wait_before_shutdown
+
+    def _format_deadlock(self, deadlock_instances: List[str]) -> str:
+        """Create and return formatted deadlock debug info.
+
+        Args:
+            deadlock_instances: list of instances waiting on eachother
+        """
+        num_instances = str(len(deadlock_instances))
+        lines = [f"The following {num_instances} instances are dead-locked:"]
```

Comment:
We have deadlocks everywhere, but here things are dead-locked. I'm not seeing any hyphens in the dictionary, and "dead-" can also be used an an adjective (e.g. "I was dead sure") so I'd go with "deadlocked" instead.

(A nit, a nit, I found a nit! *Dances happy nitpicking dance*)